# GPS-SUMO 2.0
Small ubiquitin-like modifiers (SUMOs) are tiny but important protein regulators involved in orchestrating a broad spectrum of biological processes, either by covalently modifying protein substrates, or by non-covalently interacting with other proteins. The identification of sumoylation sites and SUMO-interacting motifs (SIMs) in proteins is fundamental for better understanding SUMOs.

Here, we developed an updated online service, GPS-SUMO 2.0, using a non-redundant and non-homologous data set of 56,343 sumoylation sites in 9,731 proteins and 176 SIMs in 104 proteins. For prediction of sumoylation sites and SIMs, we integrated 11 types of sequence features and 3 machine learning algorithms, including penalized logistic regression (PLR), deep neural network (DNN), and transformer neural network (TNN). For users, one or multiple protein sequences or identifiers could be inputted, while the prediction results will be shown in a tabular list. Besides the basic statistics, we integrated the knowledge of 35 public resources to annotate sumoylation sites or SIMs, including but not limited to the experimental evidence, physical interactions, 3D structures, and disorder propensities. We believe that GPS-SUMO 2.0 can serve as a useful service for further analysis of SUMOs.
